### DISCUSSION TITLE
Add effect hadling support

### DISCUSSION THREAD

**Discussion by etheilred** [dotnet/csharplang#3488] [Language Ideas]
Look at the following example:

``` c#
void A()
        {
            try
            {
                var t = new User("Mark").Get() ?? throw new ArgumentNullException();
                Console.WriteLine(t);
            }
            catch (ArgumentNullException e)
            {
                // Handle exception
            }
        }
    }
```

Obviously, if Get() returns null, print will not be called. What can be done:

```c#
 void A()
        {
            try
            {
                var t = new User("Mark").Get() ?? perform new ArgumentNullEffect();
                Console.WriteLine(t);
            }
            handle (ArgumentNullEffect)
            {
                resume Users.Default.Get();
            }
        }
```

perform / handle idiom allows to define a common way of recovering from an error and continue execution. This is more handy than if ... else approach because you 
1. Write handler's code once
2. Don't care about errors explicitly in code - it is done in the outer block

**Comment by spydacarnage** [comment]
That, to me, looks far too much like `On Error Goto ...`
But in that example, why not just write:
```csharp
void A()
{
    var t = new User("Mark").Get();
    if (t != null)
        Console.WriteLine(t);
}
```

**Comment by svick** [comment]
I don't understand how would this feature be actually useful, but that might be just because of the example. Because in that example:

1. `new User("Mark").Get() ?? Users.Default.Get()` would be clearer and shorter.
2. Even if you were repeatedly calling `new User(name).Get()` where you wanted to substitute `Default` on failure, I think writing a helper function like `GetUserOrDefault()` would be a better solution.
3. There is seemingly no connection between `ArgumentNullEffect` and `Users`, so that code looks brittle to me: if another part of the code "performed" `ArgumentNullEffect` and you would "resume" it with `Users.Default.Get()`, that would be a bug.

Maybe you have a more convincing explanation and example, showing why this feature would be useful?
